### PR TITLE
refactor: construct token by token list

### PR DESCRIPTION
**Pull request by hanszhoou**
## Description
https://mask.atlassian.net/browse/MF-1847

Closes # (NO_ISSUE)

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx [hunk 1]
```diff
@@ -183,7 +196,16 @@ export function TraderDialog() {
...
                         <Trader
-                            {...traderProps}
+                            defaultInputCoin={
+                                defaultInputCoin
+                                    ? (inputToken as FungibleToken<ChainId, SchemaType.Native | SchemaType.ERC20>)
+                                    : undefined
+                            }
+                            defaultOutputCoin={
+                                defaultOutputCoin
+                                    ? (outputToken as FungibleToken<ChainId, SchemaType.Native | SchemaType.ERC20>)
```
Comment (hanszhoou) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx:204]: Remove Casting after Swap supports multi chains.

File: packages/web3-state/src/Token.ts [hunk 2]
```diff
@@ -175,4 +194,12 @@ export class TokenState<ChainId, SchemaType> implements Web3TokenState<ChainId,
         this.blockOrUnblockToken(address, token, 'block')
         this.addOrRemoveToken(address, token, 'add')
     }
+
+    async createFungibleToken(chainId: ChainId, address: string, token?: FungibleToken<ChainId, SchemaType>) {
+        return token
```
Comment (guanbinrui) [inline packages/web3-state/src/Token.ts]:
```suggestion
        throw new Error('Not implemented yet.')
```

File: packages/web3-state/src/Token.ts [hunk 3]
```diff
@@ -175,4 +194,12 @@ export class TokenState<ChainId, SchemaType> implements Web3TokenState<ChainId,
...
+    }
+
+    async createNonFungibleToken(chainId: ChainId, address: string, token?: NonFungibleToken<ChainId, SchemaType>) {
+        return token
```
Comment (guanbinrui) [inline packages/web3-state/src/Token.ts]:
```suggestion
        throw new Error('Not implemented yet.')
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx [hunk 4]
```diff
@@ -183,7 +194,16 @@ export function TraderDialog() {
                 <Web3ContextProvider value={{ pluginID: NetworkPluginID.PLUGIN_EVM, chainId }}>
                     <AllProviderTradeContext.Provider>
                         <Trader
-                            {...traderProps}
+                            defaultInputCoin={
+                                defaultInputCoin
+                                    ? (inputToken as FungibleToken<ChainId, SchemaType.Native | SchemaType.ERC20>)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx:199]:
```suggestion
                                    ? (inputToken as FungibleToken<ChainId, SchemaType>)
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx [hunk 5]
```diff
@@ -183,7 +194,16 @@ export function TraderDialog() {
...
+                                    : undefined
+                            }
+                            defaultOutputCoin={
+                                defaultOutputCoin
+                                    ? (outputToken as FungibleToken<ChainId, SchemaType.Native | SchemaType.ERC20>)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx:204]:
```suggestion
                                    ? (outputToken as FungibleToken<ChainId, SchemaType>)
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx [hunk 6]
```diff
@@ -93,30 +121,13 @@ export function TraderDialog() {
...
-                    : undefined
 
                 setTraderProps({
-                    defaultInputCoin: inputToken,
-                    defaultOutputCoin: outputToken,
+                    defaultInputCoin: defaultInputCoin as FungibleToken<ChainId, SchemaType.Native | SchemaType.ERC20>,
+                    defaultOutputCoin: defaultOutputCoin as FungibleToken<
+                        ChainId,
+                        SchemaType.Native | SchemaType.ERC20
+                    >,
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx:130]:
```suggestion
                    defaultInputCoin: defaultInputCoin as FungibleToken<ChainId, SchemaType>,
                    defaultOutputCoin: defaultOutputCoin as FungibleToken<
                        ChainId,
                        SchemaType
                    >,
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx [hunk 7]
```diff
@@ -79,6 +79,34 @@ export function TraderDialog() {
...
+            defaultInputCoin?.address ?? '',
+            defaultInputCoin?.name ?? '',
+            defaultInputCoin?.symbol ?? '',
+            defaultInputCoin?.decimals ?? 0,
+        ),
+    )
+    const { value: outputToken } = useFungibleToken(
+        undefined,
+        defaultOutputCoin?.address,
+        createFungibleToken<ChainId, SchemaType.Native | SchemaType.ERC20>(
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx:100]:
```suggestion
        createFungibleToken<ChainId, SchemaType>(
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx [hunk 8]
```diff
@@ -79,6 +79,34 @@ export function TraderDialog() {
 
     const [traderProps, setTraderProps] = useState<TraderProps>()
     const chainIdRef = useRef<ChainId>(chainId)
+    const defaultInputCoin = traderProps?.defaultInputCoin
+    const defaultOutputCoin = traderProps?.defaultOutputCoin
+
+    const { value: inputToken } = useFungibleToken(
+        undefined,
+        defaultInputCoin?.address,
+        createFungibleToken<ChainId, SchemaType.Native | SchemaType.ERC20>(
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx:88]:
```suggestion
        createFungibleToken<ChainId, SchemaType>(
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/TradeView.tsx [hunk 9]
```diff
@@ -26,11 +28,23 @@ export interface TradeViewProps extends withClasses<'root'> {
...
+        { chainId },
+    )
     const classes = useStylesExtends(useStyles(), props)
     return (
         <div className={classes.root}>
             <AllProviderTradeContext.Provider>
-                <Trader {...TraderProps} chainId={chainId} classes={{ root: classes.trade }} settings />
+                <Trader
+                    {...TraderProps}
+                    defaultInputCoin={inputToken as FungibleToken<ChainId, SchemaType.Native | SchemaType.ERC20>}
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/TradeView.tsx:43]:
```suggestion
                    defaultInputCoin={inputToken as FungibleToken<ChainId, SchemaType>}
```

File: packages/mask/src/extension/popups/pages/Swap/SwapBox/index.tsx [hunk 10]
```diff
@@ -1,37 +1,41 @@
...
+    )
 
     useUpdateEffect(() => {
         navigate(PopupRoutes.Swap, { replace: true })
     }, [chainId])
 
-    return <Trader defaultInputCoin={coin} chainId={chainId} />
+    return (
+        <Trader
+            defaultInputCoin={coin as FungibleToken<ChainId, SchemaType.Native | SchemaType.ERC20>}
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Swap/SwapBox/index.tsx:37]:
```suggestion
            defaultInputCoin={coin as FungibleToken<ChainId, SchemaType>}
```

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx (hunk 4 above)
Comment (hanszhoou) [reply packages/mask/src/plugins/Trader/SNSAdaptor/trader/TraderDialog.tsx:199]: Type 'FungibleToken<ChainId, SchemaType>' is not assignable to type 'FungibleToken<ChainId, SchemaType.Native | SchemaType.ERC20>'.

File: packages/web3-state/src/Token.ts (hunk 3 above)
Comment (hanszhoou) [reply packages/web3-state/src/Token.ts]: It's better to not implement `createNonFungibleToken` for `TokenState` in `packages/web3-state/src/Token.ts`

File: packages/mask/src/extension/popups/pages/Swap/SwapBox/index.tsx (hunk 10 above)
Comment (hanszhoou) [reply packages/mask/src/extension/popups/pages/Swap/SwapBox/index.tsx:37]: Type 'FungibleToken<ChainId, SchemaType>' is not assignable to type 'FungibleToken<ChainId, SchemaType.Native | SchemaType.ERC20>'.

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**